libr/util/json.c

```diff
@@ -0,0 +1,374 @@
+				}
+				int sz = r_utf8_encode ((ut8 *)d, codepoint);
+				if (!s) {
+					R_JSON_REPORT_ERROR ("invalid codepoint", ps);
+					return 0;
+				}
+				d += sz;
+				p += 5;
+				break;
+			}
+			default:
+				// leave untouched
+				*d++ = c;
+				break;
+			}
+		} else {
+			*d++ = c;
+		}
+	}
+	R_JSON_REPORT_ERROR ("no closing quote for string", s);
+	return 0;
+}
+
+static char *skip_block_comment(char *p) {
+	// assume p[-2]=='/' && p[-1]=='*'
+	char *ps = p - 2;
+	if (!*p) {
+		R_JSON_REPORT_ERROR ("endless comment", ps);
+		return 0;
+	}
+	REPEAT:
+	p = strchr (p + 1, '/');
+	if (!p) {
+		R_JSON_REPORT_ERROR ("endless comment", ps);
+		return 0;
+	}
+	if (p[-1] != '*') goto REPEAT;
+	return p + 1;
+}
+
+static char *parse_key(const char **key, char *p) {
+	// on '}' return with *p=='}'
+	char c;
+	while ((c = *p++)) {
+		if (c == '"') {
+			*key = unescape_string (p, &p);
+			if (!*key) {
+				return 0; // propagate error
+			}
+			while (*p && IS_WHITECHAR(*p)) {
+				p++;
+			}
+			if (*p == ':') {
+				return p + 1;
+			}
+			R_JSON_REPORT_ERROR ("unexpected chars", p);
+			return 0;
+		} else if (IS_WHITECHAR(c) || c == ',') {
+			// continue
+		} else if (c == '}') {
+			return p - 1;
+		} else if (c == '/') {
+			if (*p == '/') { // line comment
+				char *ps = p - 1;
+				p = strchr (p + 1, '\n');
+				if (!p) {
+					R_JSON_REPORT_ERROR ("endless comment", ps);
+					return 0; // error
+				}
+				p++;
+			} else if (*p == '*') { // block comment
+				p = skip_block_comment (p + 1);
+				if (!p) {
+					return 0;
+				}
+			} else {
+				R_JSON_REPORT_ERROR ("unexpected chars", p - 1);
+				return 0; // error
+			}
+		} else {
+			R_JSON_REPORT_ERROR ("unexpected chars", p - 1);
+			return 0; // error
+		}
+	}
+	R_JSON_REPORT_ERROR ("unexpected chars", p - 1);
+	return 0; // error
+}
+
+static char *parse_value(RJson *parent, const char *key, char *p) {
+	RJson *js;
+	while (1) {
+		switch (*p) {
+		case '\0':
+			R_JSON_REPORT_ERROR ("unexpected end of text", p);
+			return 0; // error
+		case ' ':
+		case '\t':
+		case '\n':
+		case '\r':
+		case ',':
+			// skip
+			p++;
+			break;
+		case '{':
+			js = create_json (R_JSON_OBJECT, key, parent);
+			p++;
+			while (1) {
+				const char *new_key;
+				p = parse_key (&new_key, p);
+				if (!p) {
+					return 0; // error
+				}
+				if (*p == '}') {
+					return p + 1; // end of object
+				}
+				p = parse_value (js, new_key, p);
+				if (!p) {
+					return 0; // error
+				}
+			}
+		case '[':
+			js = create_json (R_JSON_ARRAY, key, parent);
+			p++;
+			while (1) {
+				p = parse_value (js, 0, p);
+				if (!p) {
+					return 0; // error
+				}
+				if (*p == ']') {
+					return p + 1; // end of array
+				}
+			}
+		case ']':
+			return p;
+		case '"':
+			p++;
+			js = create_json (R_JSON_STRING, key, parent);
+			js->str_value = unescape_string (p, &p);
+			if (!js->str_value) {
+				return 0; // propagate error
```

Comment:
Missing rjsonreport call here. Is this on purpose?

Reply:
Yes, unescape_string does more detailed reporting about what is wrong with the string itself.